### DISCUSSION TITLE
Help implementing typing with Mypy?

### DISCUSSION THREAD

**Discussion by kuwv** [rq/rq#1847] [General]
I see typing is mostly implemented within `rq`. I can help finish this if you like? @selwin

**Comment by selwin** [comment]
Of course :)

**Comment by kuwv** [comment]
@selwin What do you think about updating `rq` build and tools to that of `pydantic`: https://github.com/pydantic/pydantic/blob/main/pyproject.toml

Move to declarative build. Implement pylint.

**Reply by ccrvlh** [reply]
Interesting they're using Hatch, I actually hadn't seen this being used before, will definitely take a look (been using Poetry myself and really like working with pyproject).

Edit: And Ruff, interesting.

**Reply by selwin** [reply]
I'm ok with transitioning to more modern tooling provided they are done gradually. Perhaps we can move to `pyproject` first and consider our next steps from there.

**Reply by kuwv** [reply]
> Interesting they're using Hatch, I actually hadn't seen this being used before, will definitely take a look (been using Poetry myself and really like working with pyproject).
> 
> Edit: And Ruff, interesting.

I recently started moving away from poetry because the versioning spec they use forces the dependencies to be separated. It's really good otherwise, just not compatible with anything else.

**Reply by ccrvlh** [reply]
I think the easiest way is to start with Flake8, Pytest and MyPy, while keeping the packaging with the setup.py and the requirements.txt.

**Reply by selwin** [reply]
> I think the easiest way is to start with Flake8, Pytest and MyPy, while keeping the packaging with the setup.py and the requirements.txt.

I'd like to use [ruff](https://beta.ruff.rs/docs/) instead of Flake8 though.

**Reply by kuwv** [reply]
I was curious about ruff and found this: https://github.com/charliermarsh/ruff/issues/689.

I think it still looks promising speed wise though if you should choose to go that route.

**Reply by selwin** [reply]
Yeah, we can still go with ruff to enforce flake8.

**Comment by kuwv** [comment]
Also, I noticed a lot of Python2 cruft leftover in `rq/local.py`.

**Reply by selwin** [reply]
It would be great if we can remove whatever's no longer relevant in this project :)

**Reply by kuwv** [reply]
Let me ask before I jump into it. Do you (or anyone) know what the concern is here: https://github.com/rq/rq/blob/master/rq/local.py#L18

The `thread` module was replaced with `threading` in python3, but  the additional try / except makes me think something else is going on here.

**Reply by selwin** [reply]
> The thread module was replaced with threading in python3, but the additional try / except makes me think something else is going on here.

Even I'm not sure what's going on there. `local.py` is something I'd like to get rid of in RQ v2 so we should move towards that and start putting in deprecation warnings whenever its used.

**Reply by ccrvlh** [reply]
> > The thread module was replaced with threading in python3, but the additional try / except makes me think something else is going on here.
> 
> 
> 
> Even I'm not sure what's going on there. `local.py` is something I'd like to get rid of in RQ v2 so we should move towards that and start putting in deprecation warnings whenever its used.

I dont have the code here, but I was working on a PR that removed `use_connection` which has a deprecation warning for 3 years now. With the context manager deprecated `with Connection()` I believe this is mostly used to pass the connection around internally - moving towards 100% explicit connection handling and we should be delete to delete this altogether.
